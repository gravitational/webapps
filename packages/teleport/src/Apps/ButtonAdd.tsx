/*
Copyright 2020 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { ButtonPrimary } from 'design';

const docUrl = 'https://gravitational.com/teleport/docs/';

export default function ButtonAdd(props: Props) {
  if (!props.isEnterprise) {
    return (
      <ButtonPrimary
        {...props}
        width="240px"
        onClick={() => null}
        as="a"
        target="_blank"
        href={docUrl}
      >
        View documentation
      </ButtonPrimary>
    );
  }

  if (props.canCreate) {
    return (
      <ButtonPrimary width="240px" {...props}>
        Add Application
      </ButtonPrimary>
    );
  }

  return null;
}

type Props = {
  isEnterprise: boolean;
  canCreate: boolean;
  onClick?: () => void;
  mb?: string;
  mx?: string;
  width?: string;
  kind?: string;
};
