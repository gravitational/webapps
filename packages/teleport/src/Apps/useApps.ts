/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { useState, useEffect } from 'react';
import { App } from 'teleport/services/apps';
import useAttempt from 'shared/hooks/useAttemptNext';
import useTeleport from 'teleport/useTeleport';
import useStickyClusterId from 'teleport/useStickyClusterId';

export default function useApps() {
  const ctx = useTeleport();
  const canCreate = ctx.storeUser.getTokenAccess().create;
  const [isAddAppVisible, setAppAddVisible] = useState(false);
  const { clusterId } = useStickyClusterId();
  const { attempt, run } = useAttempt('processing');
  const [apps, setApps] = useState([] as App[]);
  const isEnterprise = ctx.isEnterprise;

  function refresh() {
    return run(() => ctx.appService.fetchApps(clusterId).then(setApps));
  }

  const hideAddApp = () => {
    setAppAddVisible(false);
    refresh();
  };

  const showAddApp = () => {
    setAppAddVisible(true);
  };

  useEffect(() => {
    refresh();
  }, [clusterId]);

  return {
    clusterId,
    isEnterprise,
    isAddAppVisible,
    hideAddApp,
    showAddApp,
    canCreate,
    attempt,
    apps,
  };
}

export type State = ReturnType<typeof useApps>;
