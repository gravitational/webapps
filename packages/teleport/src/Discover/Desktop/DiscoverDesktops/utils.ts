import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
