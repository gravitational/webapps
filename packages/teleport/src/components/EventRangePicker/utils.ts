import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
