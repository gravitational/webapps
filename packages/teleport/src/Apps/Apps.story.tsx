/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Apps } from './Apps';
import { State } from './useApps';
import { apps } from './fixtures';

export default {
  title: 'Teleport/Apps',
  excludeStories: ['props'],
};

export const Loaded = () => <Apps {...props} />;

export const PaginationUnsupported = () => (
  <Apps
    {...props}
    results={{ ...props.results, paginationUnsupported: true }}
  />
);

export const Empty = () => (
  <Apps {...props} results={{ apps: [] }} isSearchEmpty={true} />
);

export const EmptyReadOnly = () => (
  <Apps
    {...props}
    results={{ apps: [] }}
    canCreate={false}
    isSearchEmpty={true}
  />
);

export const Loading = () => (
  <Apps {...props} attempt={{ status: 'processing' }} />
);

export const Failed = () => (
  <Apps
    {...props}
    attempt={{ status: 'failed', statusText: 'some error message' }}
  />
);

export const props: State = {
  results: {
    apps,
    totalCount: apps.length,
  },
  fetchStatus: '',
  attempt: { status: 'success' },
  clusterId: 'im-a-cluster',
  isLeafCluster: false,
  isEnterprise: false,
  isAddAppVisible: false,
  canCreate: true,
  hideAddApp: () => null,
  showAddApp: () => null,
  fetchNext: () => null,
  fetchPrev: () => null,
  pageSize: apps.length,
  from: 1,
  to: apps.length,
  params: {
    search: '',
    query: '',
    sort: { fieldName: 'name', dir: 'ASC' },
  },
  setParams: () => null,
  setSort: () => null,
  startKeys: [''],
  pathname: '',
  replaceHistory: () => null,
  isSearchEmpty: false,
  onLabelClick: () => null,
};
