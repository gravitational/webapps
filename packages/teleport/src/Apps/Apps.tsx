/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import { Danger } from 'design/Alert';
import { Indicator, Box } from 'design';
import AppList from './AppList';
import Empty from './Empty';
import AddApp from './AddApp';
import ButtonAdd from './ButtonAdd';
import useApps, { State } from './useApps';

export default function Container() {
  const state = useApps();
  return <Apps {...state} />;
}

export function Apps(props: State) {
  const {
    clusterId,
    isLeafCluster,
    isEnterprise,
    isAddAppVisible,
    showAddApp,
    hideAddApp,
    canCreate,
    attempt,
    apps,
  } = props;

  const isEmpty = attempt.status === 'success' && apps.length === 0;
  const hasApps = attempt.status === 'success' && apps.length > 0;

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center" justifyContent="space-between">
        <FeatureHeaderTitle>Applications</FeatureHeaderTitle>
        <ButtonAdd
          isLeafCluster={isLeafCluster}
          isEnterprise={isEnterprise}
          canCreate={canCreate}
          onClick={showAddApp}
        />
      </FeatureHeader>
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'failed' && <Danger>{attempt.statusText} </Danger>}
      {hasApps && <AppList apps={apps} />}
      {isEmpty && (
        <Empty
          isLeafCluster={isLeafCluster}
          isEnterprise={isEnterprise}
          clusterId={clusterId}
          canCreate={canCreate}
          onCreate={showAddApp}
        />
      )}
      {isAddAppVisible && <AddApp onClose={hideAddApp} />}
    </FeatureBox>
  );
}
