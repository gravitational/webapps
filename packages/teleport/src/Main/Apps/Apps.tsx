/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import { Danger } from 'design/Alert';
import {
  Indicator,
  Text,
  Box,
  Flex,
  ButtonPrimary,
  ButtonSecondary,
} from 'design';
import Card from 'design/Card';
import Image from 'design/Image';
import AppList from './AppList';
import useApps from './useApps';
import { emptyPng } from './assets';

const docUrl = 'https://gravitational.com/teleport/docs/';

export default function Container() {
  const state = useApps();
  return <Apps {...state} />;
}

export function Apps({ attempt, apps }: Props) {
  const isEmpty = attempt.status === 'success' && apps.length === 0;
  const hasApps = attempt.status === 'success' && !isEmpty;

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle>Applications</FeatureHeaderTitle>
        {hasApps && (
          <ButtonSecondary
            as="a"
            href={docUrl}
            target="_blank"
            ml="auto"
            width="240px"
          >
            View Documentation
          </ButtonSecondary>
        )}
      </FeatureHeader>
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'failed' && <Danger>{attempt.statusText} </Danger>}
      {isEmpty && <Empty />}
      {hasApps && <AppList apps={apps} />}
    </FeatureBox>
  );
}

const Empty = () => {
  return (
    <Card maxWidth="700px" my={4} mx="auto" p={5} as={Flex} alignItems="center">
      <Box width={4 / 10}>
        <Image src={emptyPng} width={'100%'} />
      </Box>
      <Box width={6 / 10} ml={5}>
        <Box mb={6}>
          <Text typography="h6" mb={3} caps>
            Secure Your First Application
          </Text>
          <Text typography="subtitle1" mb={3}>
            Teleport Application Access provides secure access to internal
            applications without the need for a VPN, but with the full control
            and audibility of Teleport.
          </Text>
        </Box>
        <ButtonPrimary as="a" href={docUrl} target="_blank" width="240px">
          View Quickstart Guide
        </ButtonPrimary>
      </Box>
    </Card>
  );
};

type Props = ReturnType<typeof useApps>;
