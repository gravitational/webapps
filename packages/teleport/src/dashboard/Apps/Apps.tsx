/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import { Danger } from 'design/Alert';
import {
  Indicator,
  Text,
  Box,
  Flex,
  ButtonPrimary,
  ButtonSecondary,
} from 'design';
import Card from 'design/Card';
import Image from 'design/Image';
import AppListCards from './AppListCards';
import useApps from './useApps';

const aapPng = require('./application.png');

export default function Container() {
  const state = useApps();
  return <Apps {...state} />;
}

/**
 * Apps is a view component that based on permissions:
 * - Displays list of running apps
 * - Allows user to view details of an app
 */
export function Apps({ attempt, apps }: Props) {
  const isEmpty = attempt.status === 'success' && apps.length === 0;
  const hasApps = attempt.status === 'success' && !isEmpty;

  function onClickDoc() {
    window.open('https://gravitational.com/teleport/docs/', '_blank');
  }

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle>Applications</FeatureHeaderTitle>
        {hasApps && (
          <ButtonSecondary onClick={onClickDoc} ml="auto" width="240px">
            View Documentation
          </ButtonSecondary>
        )}
      </FeatureHeader>
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'failed' && <Danger>{attempt.statusText} </Danger>}
      {isEmpty && <Empty onClick={onClickDoc} />}
      {hasApps && <AppListCards apps={apps} />}
    </FeatureBox>
  );
}

const Empty = ({ onClick }) => {
  return (
    <Card maxWidth="700px" my={4} mx="auto" p={5} as={Flex} alignItems="center">
      <Box width={4 / 10}>
        <Image src={aapPng.default} width={'100%'} />
      </Box>
      <Box width={6 / 10} ml={5}>
        <Box mb={6}>
          <Text typography="h6" mb={3} caps>
            Secure Your First Application
          </Text>
          <Text typography="subtitle1" mb={3}>
            Teleport Application Access provides secure access to internal
            applications without the need for a VPN, but with the full control
            and audibility of Teleport.
          </Text>
        </Box>
        <ButtonPrimary onClick={onClick} width="240px">
          View Quickstart Guide
        </ButtonPrimary>
      </Box>
    </Card>
  );
};

type Props = ReturnType<typeof useApps>;
