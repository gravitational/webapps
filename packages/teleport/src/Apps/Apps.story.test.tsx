import React from 'react';
import { render } from 'design/utils/testing';
import { Loaded, Failed, Empty, EmptyReadOnly } from './Apps.story';

jest.mock('teleport/useStickyClusterId', () =>
  jest.fn(() => ({ clusterId: 'im-a-cluster', isLeafCluster: false }))
);

test('loaded state', async () => {
  const { container, findAllByText } = render(<Loaded />);
  await findAllByText(/Applications/i);

  expect(container).toMatchSnapshot();
});

test('failed state', async () => {
  const { container, findAllByText } = render(<Failed />);
  await findAllByText(/some error message/i);

  expect(container).toMatchSnapshot();
});

test('empty state for enterprise, can create', () => {
  const { container } = render(<Empty />);
  expect(container).toMatchSnapshot();
});

test('readonly empty state', () => {
  const { container } = render(<EmptyReadOnly />);
  expect(container).toMatchSnapshot();
});
