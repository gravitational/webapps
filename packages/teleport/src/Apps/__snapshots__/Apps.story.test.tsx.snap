// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty state 1`] = `
.c6 {
  box-sizing: border-box;
  margin-left: 24px;
  margin-right: 24px;
}

.c8 {
  box-sizing: border-box;
}

.c9 {
  box-sizing: border-box;
  margin-bottom: 40px;
  padding-right: 24px;
}

.c3 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  width: 240px;
}

.c3:active {
  opacity: 0.56;
}

.c3:hover,
.c3:focus {
  background: #651FFF;
}

.c3:active {
  background: #354AA4;
}

.c3:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c12 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  margin-bottom: 8px;
  margin-left: auto;
  margin-right: auto;
  width: 240px;
}

.c12:active {
  opacity: 0.56;
}

.c12:hover,
.c12:focus {
  background: #651FFF;
}

.c12:active {
  background: #354AA4;
}

.c12:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c5 {
  box-sizing: border-box;
  max-width: 700px;
  margin-left: auto;
  margin-right: auto;
  padding-top: 24px;
  padding-bottom: 24px;
  background-color: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c7 {
  display: block;
  outline: none;
  width: 180px;
}

.c10 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  font-size: 14px;
  line-height: 24px;
  margin: 0px;
  margin-bottom: 16px;
}

.c11 {
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0px;
  margin-bottom: 16px;
}

.c4 {
  box-sizing: border-box;
  max-width: 700px;
  margin-left: auto;
  margin-right: auto;
  padding-top: 24px;
  padding-bottom: 24px;
  background-color: #222C59;
  display: flex;
  align-items: center;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  flex-shrink: 0;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Applications
      </div>
      <button
        class="c3"
        kind="primary"
        title="Add an application to the root cluster"
        width="240px"
      >
        Add Application
      </button>
    </div>
    <div
      class="c4 sc-AxirZ c5"
    >
      <div
        class="c6"
      >
        <img
          class="c7"
          src="file_stub"
          width="180px"
        />
      </div>
      <div
        class="c8"
      >
        <div
          class="c9"
        >
          <div
            class="c10"
          >
            SECURE YOUR FIRST APPLICATION
          </div>
          <div
            class="c11"
          >
            Teleport Application Access provides secure access to internal applications without the need for a VPN and with the auditability and control of Teleport.
          </div>
        </div>
        <button
          class="c12"
          kind="primary"
          title="Add an application to the root cluster"
          width="240px"
        >
          Add Application
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`loaded state 1`] = `
.c3 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  width: 240px;
}

.c3:active {
  opacity: 0.56;
}

.c3:hover,
.c3:focus {
  background: #651FFF;
}

.c3:active {
  background: #354AA4;
}

.c3:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c9 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c11 {
  display: inline-block;
  transition: color .3s;
  margin-bottom: 16px;
  color: #FFFFFF;
  font-size: 62px;
}

.c7 {
  align-items: center;
  border: none;
  cursor: pointer;
  display: flex;
  outline: none;
  border-radius: 50%;
  overflow: visible;
  justify-content: center;
  text-align: center;
  flex: 0 0 auto;
  background: transparent;
  color: inherit;
  transition: all .3s;
  -webkit-font-smoothing: antialiased;
  font-size: 16px;
  height: 32px;
  width: 32px;
}

.c7 .c8 {
  color: inherit;
}

.c7:disabled {
  color: rgba(255,255,255,0.3);
}

.c7:disabled {
  color: rgba(255,255,255,0.3);
}

.c7:hover,
.c7:focus {
  background: rgba(255,255,255,0.1);
}

.c12 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  margin: 0px;
  margin-bottom: 8px;
}

.c4 {
  box-sizing: border-box;
  display: flex;
  flex-wrap: wrap;
}

.c6 {
  box-sizing: border-box;
  width: 100%;
  display: flex;
  justify-content: center;
}

.c10 {
  box-sizing: border-box;
  padding-left: 8px;
  padding-right: 8px;
  width: 220px;
  color: rgba(255,255,255,0.87);
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  flex-shrink: 0;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

.c5 {
  box-sizing: border-box;
  margin-bottom: 24px;
  margin-right: 32px;
  height: 240px;
  width: 240px;
  background-color: #222C59;
  border-radius: 8px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  position: relative;
  box-shadow: 0 4px 32px rgba(0,0,0,0.24);
  cursor: pointer;
  transition: 0.3s;
  border: 2px solid transparent;
}

.c5:hover {
  border: 2px solid #512FC9;
  background: #2C3A73;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Applications
      </div>
      <button
        class="c3"
        kind="primary"
        title="Add an application to the root cluster"
        width="240px"
      >
        Add Application
      </button>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
        height="240px"
        width="240px"
      >
        <div
          class="c6"
          width="100%"
        >
          <button
            class="c7"
            data-testid="button"
            style="right: 10px; position: absolute; top: 10px;"
          >
            <span
              class="c8 c9 icon icon-ellipsis "
              color="light"
            />
          </button>
        </div>
        <a
          class="c10"
          color="text.primary"
          href="/web/launch/jenkins.one/one/jenkins"
          style="text-decoration: none;"
          tabindex="-1"
          target="_blank"
          width="220px"
        >
          <span
            class="c8 c11 icon icon-new-tab "
            color="light"
            font-size="62px"
          />
          <div
            class="c12"
            style="text-align: center; width: 100%;"
          >
            jenkins
          </div>
        </a>
      </div>
      <div
        class="c5"
        height="240px"
        width="240px"
      >
        <div
          class="c6"
          width="100%"
        >
          <button
            class="c7"
            data-testid="button"
            style="right: 10px; position: absolute; top: 10px;"
          >
            <span
              class="c8 c9 icon icon-ellipsis "
              color="light"
            />
          </button>
        </div>
        <a
          class="c10"
          color="text.primary"
          href="/web/launch/jenkins.two/two/jenkins"
          style="text-decoration: none;"
          tabindex="-1"
          target="_blank"
          width="220px"
        >
          <span
            class="c8 c11 icon icon-new-tab "
            color="light"
            font-size="62px"
          />
          <div
            class="c12"
            style="text-align: center; width: 100%;"
          >
            jenkins
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
